### PR TITLE
[AIRFLOW-4423] Improve date handling in mysql to gcs operator.

### PR DESCRIPTION
**Pull request by jmcarp**
* Handle TIME columns
* Ensure DATETIME and TIMESTAMP columns treated as UTC

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4423
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

* Handle TIME columns
* Ensure DATETIME and TIMESTAMP columns treated as UTC

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/operators/mysql_to_gcs.py [hunk 1]
```diff
@@ -265,27 +288,31 @@ def _upload_to_gcs(self, files_to_upload):
...
-    def _convert_types(schema, col_type_dict, row):
+    @classmethod
+    def _convert_types(cls, schema, col_type_dict, row):
+        return [
+            cls._convert_type(value, col_type_dict.get(name))
+            for name, value in zip(schema, row)
+        ]
+
+    @classmethod
+    def _convert_type(cls, value, schema_type):
```
Comment (feluelle) [inline airflow/contrib/operators/mysql_to_gcs.py:306]: The documentation for this conversion is good, but in my opinion we should always document the parameters as well + to have a cleaner sphinx generated documentation. :)

File: tests/contrib/operators/test_mysql_to_gcs_operator.py [hunk 2]
```diff
@@ -106,7 +121,7 @@ def _assert_upload(bucket, obj, tmp_filename, mime_type=None):
         op.execute(None)
 
         mysql_hook_mock_class.assert_called_once_with(mysql_conn_id=MYSQL_CONN_ID)
-        mysql_hook_mock.get_conn().cursor().execute.assert_called_once_with(SQL)
+        mysql_hook_mock.get_conn().cursor().execute.assert_called_with(SQL)
```
Comment (feluelle) [inline tests/contrib/operators/test_mysql_to_gcs_operator.py]: Please test that you executed the `tz_query` before you execute the actual `sql` statement. You can test that via [assert_has_calls](https://docs.python.org/3.6/library/unittest.mock.html#unittest.mock.Mock.assert_has_calls). Also see this https://github.com/apache/airflow/pull/5101#discussion_r275438792

**Review by feluelle — CHANGES_REQUESTED** [review]
Really nice and clean (readable / understandable) solution I think :)

I only have a few comments related to the documentation and tests.

**Comment by RosterIn** [comment]
My previous PR is somehow related as this one also address TIME field. I suggest going through the comments of the mysql lib commiter which kindly gave input:
https://github.com/apache/airflow/pull/4802

**Comment by jmcarp** [comment]
@feluelle: I updated the tests and docs like you suggested.

**Review by feluelle — APPROVED** [review]
LGTM now 👍

**Comment by jmcarp** [comment]
cc @kaxil @Fokko @XD-DENG if you have time to review

**Comment by jmcarp** [comment]
Ping @kaxil @Fokko @XD-DENG when you have time to review. Or maybe @ashb or @potiuk?

**Review by potiuk — CHANGES_REQUESTED** [review]
LGTM but I think we should also update UPDATING.md as it's backwards incompatible.

File: airflow/contrib/operators/mysql_to_gcs.py [hunk 3]
```diff
@@ -138,6 +156,11 @@ def _query_mysql(self):
         mysql = MySqlHook(mysql_conn_id=self.mysql_conn_id)
         conn = mysql.get_conn()
         cursor = conn.cursor()
+        # Ensure TIMESTAMP results are in UTC
+        tz_query = "SET time_zone = '+00:00'"
```
Comment (ashb) [inline airflow/contrib/operators/mysql_to_gcs.py]: Is it worth making this an argument to the op?

**Comment by jmcarp** [comment]
Thanks @potiuk and @ashb. This is technically a breaking change, but I think the current behavior is also incorrect for the use case of exporting to bigquery, which seems to be the main purpose of this operator, since bigquery assumes utc for timestamps without timezones. I made a note in `UPDATING.md` and added a flag to preserve the current behavior in case anybody is depending on it.

**Review by potiuk — APPROVED** [review]

File: airflow/contrib/operators/mysql_to_gcs.py (hunk 3 above)
Comment (potiuk) [reply airflow/contrib/operators/mysql_to_gcs.py]: I think this is resolved now :).

**Comment by jmcarp** [comment]
Conflicts resolved. OK to merge @potiuk @ashb ?

**Comment by ashb** [comment]
Looks good to me for master. What would you say to changing/leaving the default of the new param as `False` on 1.10 branch?

**Comment by jmcarp** [comment]
Updated the new flag to default to `False` for now @ashb.

**Review by Fokko — APPROVED** [review]

**Comment by jmcarp** [comment]
Conflicts resolved.

**Comment by Fokko** [comment]
Thanks @jmcarp

**Pull request merged**